File: python/mirage/resource/slack/slack.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.slack import SlackAccessor
from mirage.core.slack.glob import resolve_glob as _resolve_glob
from mirage.resource.base import BaseResource
from mirage.resource.slack.config import SlackConfig
from mirage.resource.slack.prompt import PROMPT, WRITE_PROMPT
from mirage.types import ResourceName


class SlackResource(BaseResource):

    accessor: SlackAccessor
    name: str = ResourceName.SLACK
    caches_reads: bool = True
    PROMPT: str = PROMPT
    WRITE_PROMPT: str = WRITE_PROMPT

    def __init__(self, config: SlackConfig) -> None:
        super().__init__()
        self.config = config
        self.accessor = SlackAccessor(self.config)
        from mirage.commands.builtin.slack import COMMANDS
        from mirage.ops.slack import OPS as SLACK_VFS_OPS

        for fn in COMMANDS:
            self.register(fn)
        for fn in SLACK_VFS_OPS:
            self.register_op(fn)

    async def resolve_glob(self, paths, prefix: str = ""):
        return await _resolve_glob(self.accessor, paths, index=self._index)

    def get_state(self) -> dict:
        return self.config_state(self.config)

    def load_state(self, state: dict) -> None:
        pass

File: typescript/packages/browser/src/resource/slack/slack.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { mountKey, mountPrefixOf } from '@struktoai/mirage-core'
import {
  BrowserSlackTransport,
  type FileStat,
  type IndexCacheStore,
  PathSpec,
  RAMIndexCacheStore,
  type RegisteredCommand,
  type RegisteredOp,
  type Resource,
  ResourceName,
  makeResolveGlob,
  SLACK_COMMANDS,
  SLACK_PROMPT,
  SLACK_VFS_OPS,
  SLACK_WRITE_PROMPT,
  SlackAccessor,
  slackRead,
  slackReaddir,
  slackStat,
} from '@struktoai/mirage-core'
import { redactSlackConfig, type SlackConfig, type SlackConfigRedacted } from './config.ts'

const resolveSlackGlob = makeResolveGlob(slackReaddir)

export interface SlackResourceState {
  type: string
  config: SlackConfigRedacted
}

export class SlackResource implements Resource {
  readonly kind: string = ResourceName.SLACK
  readonly cachesReads: boolean = true
  readonly indexTtl: number = 600
  readonly prompt: string = SLACK_PROMPT
  readonly writePrompt: string = SLACK_WRITE_PROMPT
  readonly config: SlackConfig
  readonly accessor: SlackAccessor
  readonly index: IndexCacheStore

  constructor(config: SlackConfig) {
    this.config = config
    this.accessor = new SlackAccessor(
      new BrowserSlackTransport({
        proxyUrl: config.proxyUrl,
        ...(config.getHeaders !== undefined ? { getHeaders: config.getHeaders } : {}),
      }),
    )
    this.index = new RAMIndexCacheStore({ ttl: this.indexTtl })
  }

  open(): Promise<void> {
    return Promise.resolve()
  }

  close(): Promise<void> {
    return Promise.resolve()
  }

  commands(): readonly RegisteredCommand[] {
    return SLACK_COMMANDS
  }

  ops(): readonly RegisteredOp[] {
    return SLACK_VFS_OPS
  }

  readFile(p: PathSpec): Promise<Uint8Array> {
    return slackRead(this.accessor, p, this.index)
  }

  readdir(p: PathSpec): Promise<string[]> {
    return slackReaddir(this.accessor, p, this.index)
  }

  stat(p: PathSpec): Promise<FileStat> {
    return slackStat(this.accessor, p, this.index)
  }

  glob(paths: readonly PathSpec[], prefix = ''): Promise<PathSpec[]> {
    const effective =
      prefix !== ''
        ? paths.map((p) =>
            mountPrefixOf(p.virtual, p.resourcePath) !== ''
              ? p
              : new PathSpec({
                  virtual: p.virtual,
                  directory: p.directory,
                  ...(p.pattern !== null ? { pattern: p.pattern } : {}),
                  resolved: p.resolved,
                  resourcePath: mountKey(p.virtual, prefix),
                }),
          )
        : paths
    return resolveSlackGlob(this.accessor, effective, this.index)
  }

  getState(): Promise<SlackResourceState> {
    return Promise.resolve({
      type: this.kind,
      config: redactSlackConfig(this.config),
    })
  }

  loadState(_state: SlackResourceState): Promise<void> {
    return Promise.resolve()
  }
}
